renderdoc/driver/gl/wrappers/gl_external_texture.cpp

```diff
@@ -0,0 +1,169 @@
+
+rdcarray<byte> WrappedOpenGL::GetExternalTextureData(GLuint texture)
+{
+  rdcarray<byte> pixels;
+  GLuint prevTex = 0;
+  GL.glGetIntegerv(eGL_TEXTURE_BINDING_EXTERNAL_OES, (GLint *)&prevTex);
+  GL.glBindTexture(eGL_TEXTURE_EXTERNAL_OES, texture);
+
+  GLint width = 0, height = 0;
+  GLenum internalFormat = eGL_NONE;
+  GL.glGetTexLevelParameteriv(eGL_TEXTURE_EXTERNAL_OES, 0, eGL_TEXTURE_WIDTH, &width);
+  GL.glGetTexLevelParameteriv(eGL_TEXTURE_EXTERNAL_OES, 0, eGL_TEXTURE_HEIGHT, &height);
+  GL.glGetTexLevelParameteriv(eGL_TEXTURE_EXTERNAL_OES, 0, eGL_TEXTURE_INTERNAL_FORMAT,
+                              (GLint *)&internalFormat);
+  GL.glBindTexture(eGL_TEXTURE_EXTERNAL_OES, prevTex);
+
+  size_t size = GetByteSize(width, height, 1, GetBaseFormat(internalFormat), eGL_UNSIGNED_BYTE);
+
+  pixels.resize(size);
+
+  // read pixels. ref: https://developer.arm.com/documentation/ka004859/1-0
+  GLuint prevReadFramebuffer = 0, prevPixelPackBuffer = 0, fb = 0;
+  GL.glGetIntegerv(eGL_PIXEL_PACK_BUFFER_BINDING, (GLint *)&prevPixelPackBuffer);
+  GL.glBindBuffer(eGL_PIXEL_PACK_BUFFER, 0);
+  GL.glGenFramebuffers(1, &fb);
+  GL.glGetIntegerv(eGL_READ_FRAMEBUFFER_BINDING, (GLint *)&prevReadFramebuffer);
+  GL.glBindFramebuffer(eGL_READ_FRAMEBUFFER, fb);
+  GL.glFramebufferTexture2D(eGL_READ_FRAMEBUFFER, eGL_COLOR_ATTACHMENT0, eGL_TEXTURE_EXTERNAL_OES,
+                            texture, 0);
+
+  GLint prevPixelPackAlign = 0;
+  GL.glGetIntegerv(eGL_PACK_ALIGNMENT, &prevPixelPackAlign);
+  GL.glPixelStorei(eGL_PACK_ALIGNMENT, 1);
+  GL.glReadPixels(0, 0, width, height, GetBaseFormat(internalFormat), eGL_UNSIGNED_BYTE,
+                  pixels.data());
+  GL.glFinish();
+  GL.glPixelStorei(eGL_PACK_ALIGNMENT, prevPixelPackAlign);
+  GL.glBindFramebuffer(eGL_READ_FRAMEBUFFER, prevReadFramebuffer);
+  GL.glDeleteFramebuffers(1, &fb);
+  GL.glBindBuffer(eGL_PIXEL_PACK_BUFFER, prevPixelPackBuffer);
+
+  return pixels;
+}
+
+GLeglImageOES WrappedOpenGL::CreateEGLImage(GLint width, GLint height, GLenum internalFormat,
+                                            const byte *pixels, uint64_t size)
+{
+  GLeglImageOES image = NULL;
+
+#if defined(RENDERDOC_PLATFORM_ANDROID)
+  uint32_t bufferFormat = 0;
+  switch(internalFormat)
+  {
+    case eGL_RGB8: bufferFormat = AHARDWAREBUFFER_FORMAT_R8G8B8_UNORM; break;
+    case eGL_RGBA8: bufferFormat = AHARDWAREBUFFER_FORMAT_R8G8B8A8_UNORM; break;
+    default: RDCERR("Unsupported internal format 0x%X", internalFormat);
+  }
+  AHardwareBuffer *hardwareBuffer = NULL;
+  EGLClientBuffer clientBuffer = NULL;
+
+  AHardwareBuffer_Desc hardwareBufferDesc{};
+  hardwareBufferDesc.width = width;
+  hardwareBufferDesc.height = height;
+  hardwareBufferDesc.format = bufferFormat;
+  hardwareBufferDesc.layers = 1;
+  hardwareBufferDesc.usage =
+      AHARDWAREBUFFER_USAGE_CPU_WRITE_RARELY | AHARDWAREBUFFER_USAGE_GPU_SAMPLED_IMAGE;
+
+  int res = AHardwareBuffer_allocate(&hardwareBufferDesc, &hardwareBuffer);
```

Comment:
The replay should not be based around creating actual external textures. That's outside the scope of what RenderDoc does and exposes a potential unknown API hole with creating a compatible external texture.

If you look at `gl_interop_funcs.cpp` for how the other external sharing GL APIs are implemented, you'll see that e.g. in `WrappedOpenGL::Serialise_wglDXRegisterObjectNV` for windows or `WrappedOpenGL::Serialise_glTextureStorageMem1DEXT` which is platform agnostic, the textures are  created just as plain textures but with properties to mimic the original imported texture.

The replay of this extension should function the same, not depending on any native APIs like this but during capture snapshotting their properties and on replay creating a "fake" standin for the external texture. This then means RenderDoc only needs to cover the API that is in scope (GL ES in this case) and means it behaves more predictably and understandably.